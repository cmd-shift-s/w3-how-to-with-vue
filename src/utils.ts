import MarkdownIt from 'markdown-it'
import hljs from 'highlight.js'
import 'highlight.js/styles/vs.css'

export const md = new MarkdownIt({
  highlight (code, lang) {
    return hljs.highlight(lang, code).value
  }
})
