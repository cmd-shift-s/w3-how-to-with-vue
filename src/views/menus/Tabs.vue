<template lang="html">
  <page
    title="Tabs"
    intro="tabs with CSS and JavaScript">

    <template slot="header">
    <h2>Tabs</h2>
    <p>Tabs are perfect for single page web applications, or for web pages capable of displaying different subjects:</p>
    </template>

    <sample>
      <tabs
        is-animated
        :links="['London', 'Paris', 'Tokyo']">
        <div id="London" class="tab-content">
          <h3>London</h3>
          <p>London is the capital city of England.</p>
        </div>
        <div id="Paris" class="tab-content">
          <h3>Paris</h3>
          <p>Paris is the capital of France.</p>
        </div>
        <div id="Tokyo" class="tab-content">
          <h3>Tokyo</h3>
          <p>Tokyo is the capital of Japan.</p>
        </div>
      </tabs>
    </sample>

    <template slot="examples">
    <hr>
    <h2 class="title">Create Toggleable Tabs</h2>
    <step no="1" html></step>
    <example-code :code="examples[0]"></example-code>
    <p>Create buttons to open specific tab content. All &lt;div&gt; elements with <code>class="tabcontent"</code> are hidden by default (with CSS &amp; JS). When the user clicks on a button - it will open the tab content that "matches" this button.</p>
    <hr>
    <step no="2" css>
      <p>Style the buttons and the tab content:</p>
    </step>
    <example-code :code="examples[1]"></example-code>
    <step no="3" js></step>
    <example-code :code="examples[2]"></example-code>
    <hr>
    <h2 class="title">Fade in Tabs:</h2>
    <p>If you want to fade in the tab content, add the following CSS:</p>
    <example-code :code="examples[3]"></example-code>
    <hr>
    <h2 class="title">Show a tab by default</h2>
    <p>To open a specific tab on page load, use JavaScript to "click" on the specified tab button:</p>
    <example-code :code="examples[4]"></example-code>
    <hr>
    <h2 class="title">Close a tab</h2>
    <p>If you want to close a specific tab, use JavaScript to hide the tab with a click of a button:</p>
    <example-code :code="examples[5]"></example-code>
    </template>
  </page>
</template>

<script>
import Tabs from '@/components/menus/Tabs.vue'

export default {
  name: 'tabs-page',
  data() {
    return {
      examples: [
        [
          '``` html',
          '<div class="tab">',
          '  <button class="tablinks" onclick="openCity(event, \'London\')">London</button>',
          '  <button class="tablinks" onclick="openCity(event, \'Paris\')">Paris</button>',
          '  <button class="tablinks" onclick="openCity(event, \'Tokyo\')">Tokyo</button>',
          '</div>',
          '',
          '<div id="London" class="tabcontent">',
          '  <h3>London</h3>',
          '  <p>London is the capital city of England.</p>',
          '</div>',
          '',
          '<div id="Paris" class="tabcontent">',
          '  <h3>Paris</h3>',
          '  <p>Paris is the capital of France.</p> ',
          '</div>',
          '',
          '<div id="Tokyo" class="tabcontent">',
          '  <h3>Tokyo</h3>',
          '  <p>Tokyo is the capital of Japan.</p>',
          '</div>',
          '```'
        ],
        [
          '``` css',
          '/* Style the tab */',
          'div.tab {',
          '    overflow: hidden;',
          '    border: 1px solid #ccc;',
          '    background-color: #f1f1f1;',
          '}',
          '',
          '/* Style the buttons inside the tab */',
          'div.tab button {',
          '    background-color: inherit;',
          '    float: left;',
          '    border: none;',
          '    outline: none;',
          '    cursor: pointer;',
          '    padding: 14px 16px;',
          '    transition: 0.3s;',
          '}',
          '',
          '/* Change background color of buttons on hover */',
          'div.tab button:hover {',
          '    background-color: #ddd;',
          '}',
          '',
          '/* Create an active/current tablink class */',
          'div.tab button.active {',
          '    background-color: #ccc;',
          '}',
          '',
          '/* Style the tab content */',
          '.tabcontent {',
          '    display: none;',
          '    padding: 6px 12px;',
          '    border: 1px solid #ccc;',
          '    border-top: none;',
          '}',
          '```'
        ],
        [
          '``` js',
          'function openCity(evt, cityName) {',
          '    // Declare all variables',
          '    var i, tabcontent, tablinks;',
          '',
          '    // Get all elements with class="tabcontent" and hide them',
          '    tabcontent = document.getElementsByClassName("tabcontent");',
          '    for (i = 0; i < tabcontent.length; i++) {',
          '        tabcontent[i].style.display = "none";',
          '    }',
          '',
          '    // Get all elements with class="tablinks" and remove the class "active"',
          '    tablinks = document.getElementsByClassName("tablinks");',
          '    for (i = 0; i < tablinks.length; i++) {',
          '        tablinks[i].className = tablinks[i].className.replace(" active", "");',
          '    }',
          '',
          '    // Show the current tab, and add an "active" class to the button that opened the tab',
          '    document.getElementById(cityName).style.display = "block";',
          '    evt.currentTarget.className += " active";',
          '}',
          '```'
        ],
        [
          '``` css',
          '.tabcontent {',
          '    -webkit-animation: fadeEffect 1s;',
          '    animation: fadeEffect 1s; /* Fading effect takes 1 second */',
          '}',
          '',
          '@-webkit-keyframes fadeEffect {',
          '    from {opacity: 0;}',
          '    to {opacity: 1;}',
          '}',
          '',
          '@keyframes fadeEffect {',
          '    from {opacity: 0;}',
          '    to {opacity: 1;}',
          '}',
          '```'
        ],
        [
          '``` html',
          '<button class="tablinks" onclick="openCity(event, \'London\')" id="defaultOpen">London</button>',
          '',
          '<script>',
          '// Get the element with id="defaultOpen" and click on it',
          'document.getElementById("defaultOpen").click();',
          '</' + 'script>',
          '```'
        ],
        [
          '``` html',
          '<!-- Click on the <span> element to close the tab -->',
          '',
          '<div id="London" class="tabcontent">',
          '  <h3>London</h3>',
          '  <p>London is the capital city of England.</p>',
          '  <span onclick="this.parentElement.style.display=\'none\'">x</span> ',
          '</div>',
          '```'
        ]
      ]
    }
  },
  components: {
    Tabs
  }
}
</script>

<style lang="scss">
</style>
