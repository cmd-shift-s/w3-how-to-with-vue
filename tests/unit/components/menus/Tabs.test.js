import { mount, renderer } from '@/tests/unit/utils'
import Tabs from '@/components/menus/Tabs.vue'

describe('Tabs.vue', () => {
  it('render default', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test']
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-top'])
    expect(wrap.vm.currentLink).toEqual('test')

    wrap.update()

    expect(wrap.find('.tab-link').classes()).toContain('is-active')

    // tab-content의 is-active는 renderer에서 상태 값이 없기 때문에 볼 수 없다.
    const $test = wrap.find('#test')
    expect($test.classes()).toContains(['tab-content', 'is-active'])
    expect($test.text()).toEqual('test content')

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('tab-link@click', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test1', 'test2']
      },
      slots: {
        default: `
          <div>
            <div id="test1" class="tab-content">test1 content</div>
            <div id="test2" class="tab-content">test2 content</div>
          </div>
        `
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContain('tabs')
    expect(wrap.vm.currentLink).toEqual('test1')

    wrap.update()

    // 첫번째 버튼
    const $btn1 = wrap.find('.tab-link')

    expect($btn1.classes()).toContain('is-active')

    const $test1 = wrap.find('#test1')
    expect($test1.classes()).toContains(['tab-content', 'is-active'])
    expect($test1.text()).toEqual('test1 content')

    const $btn2 = wrap.find('.tab-link:nth-child(2)')
    expect($btn2.classes()).not.toContain('is-active')

    // 2번째 탭 클릭
    $btn2.trigger('click')
    wrap.update()

    expect(wrap.vm.currentLink).toEqual('test2')

    expect($btn2.classes()).toContain('is-active')

    const $test2 = wrap.find('#test2')
    expect($test2.classes()).toContains(['tab-content', 'is-active'])
    expect($test2.text()).toEqual('test2 content')

    // 이전에 선택된 tab의 is-active가 사라짐
    expect(wrap.find('.tab-link').classes()).not.toContain('is-active')
    expect(wrap.find('#test1').classes()).not.toContain('is-active')

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#is-animated', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test'],
        isAnimated: true
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-animated'])

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#position:top', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test'],
        position: 'top'
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-top'])

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#position:right', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test'],
        position: 'right'
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-right'])

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#position:bottom', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test'],
        position: 'bottom'
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-bottom'])

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('props#position:left', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test'],
        position: 'left'
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.classes()).toContains(['tabs', 'is-left'])

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })
})
