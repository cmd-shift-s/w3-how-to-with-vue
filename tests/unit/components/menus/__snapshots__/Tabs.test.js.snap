// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.vue props#is-animated 1`] = `
<div class="tabs is-animated is-top">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue props#position:bottom 1`] = `
<div class="tabs is-bottom">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue props#position:left 1`] = `
<div class="tabs is-left">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue props#position:right 1`] = `
<div class="tabs is-right">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue props#position:top 1`] = `
<div class="tabs is-top">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue render default 1`] = `
<div class="tabs is-top">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue tab-link@click 1`] = `
<div class="tabs is-top">
    <div class="tabs-links"><button class="tab-link">test1</button><button class="tab-link is-active">test2</button></div>
    <div class="tabs-contents">
        <div>
            <div id="test1" class="tab-content">test1 content</div>
            <div id="test2" class="tab-content">test2 content</div>
        </div>
    </div>
</div>
`;
